Middleware/BacktraceDebugDataHolder.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Doctrine\Bundle\DoctrineBundle\Middleware;
+
+use Symfony\Bridge\Doctrine\Middleware\Debug\DebugDataHolder;
+use Symfony\Bridge\Doctrine\Middleware\Debug\Query;
+
+use function array_merge;
+use function array_slice;
+use function debug_backtrace;
+use function in_array;
+
+use const DEBUG_BACKTRACE_IGNORE_ARGS;
+
+/** @psalm-suppress UndefinedClass */
+class BacktraceDebugDataHolder extends DebugDataHolder
+{
+    /** @var string[] */
+    private $connWithBacktraces;
+
+    /** @var mixed[][][] */
+    private $backtraces = [];
+
+    /** @param string[] $connWithBacktraces */
+    public function __construct(array $connWithBacktraces)
+    {
+        $this->connWithBacktraces = $connWithBacktraces;
+    }
+
+    public function reset(): void
+    {
+        parent::reset();
+
+        $this->backtraces = [];
+    }
+
+    public function addQuery(string $connectionName, Query $query): void
+    {
+        parent::addQuery($connectionName, $query);
+
+        // array_slice to skip middleware calls in the trace
+        if (! in_array($connectionName, $this->connWithBacktraces, true)) {
+            return;
+        }
+
+        $this->backtraces[$connectionName][] = array_slice(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS), 2);
+    }
+
+    /** @return mixed[][][] */
```

Comment:
this is actually `array<string, mixed[][]>`

Reply:
Actually, it seems it's `array<string, array<string, mixed>[]>`